include/glow/Optimizer/GraphOptimizer/NodeSplitting.h

```diff
@@ -0,0 +1,283 @@
+/**
+ * Copyright (c) Glow Contributors. See CONTRIBUTORS file.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_OPTIMIZER_GRAPHOPTIMIZER_NODESPLITTING_H
+#define GLOW_OPTIMIZER_GRAPHOPTIMIZER_NODESPLITTING_H
+
+#include "glow/Base/Type.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/Node.h"
+#include "glow/Support/Error.h"
+
+#include "llvm/ADT/ArrayRef.h"
+
+#include <functional>
+
+namespace glow {
+
+///===---------------------------------------------------------------------===//
+///                               SplitNodeOption
+///===---------------------------------------------------------------------===//
+/// Generic split node option as base type. All the split options refer to the
+/// split procedure performed with respect to the tensor associated with the
+/// output operand of a node. If a node has multiple output operands then the
+/// option commonly refers to the first output operand.
+class SplitNodeOption {
+
+  /// All the dimensions of a tensor used for splitting in increasing order.
+  std::vector<size_t> splitDims_;
+
+public:
+  /// Ctor.
+  SplitNodeOption(llvm::ArrayRef<size_t> splitDims) : splitDims_(splitDims) {}
+
+  /// \returns the split dims.
+  llvm::ArrayRef<size_t> getSplitDims() const { return splitDims_; }
+
+  /// \returns the index of \p splitDim in the \ref splitDims_ array.
+  size_t getSplitDimIdx(dim_t splitDim) const;
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  virtual std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const = 0;
```

Comment:
Should the return type be a mutable vector? If not, return an ArrayRef.

Reply:
Stupid question: if inside this function I use a local `std::vector` and return it as a `llvm::ArrayRef` which you said is nothing but a reference, wouldn't the `ArrayRef` remain dangling after the local `std::vector` is destroyed? You have to bare with me cuz I'm not an experienced C++ programmer ...

Reply:
If you create a new vector inside the function you cannot use ArrayRef to return it. You really need to return std::vector.

Alternatively, you pass an output `llvm::SmallVectorImpl<size_t> &out` parameter to the function and add your elements to this vector. In this case the function does not need to return anything as it is done through this out parameter.

Reply:
Ok it makes sense now. So for these functions I will preserve `std::vector` since these generate new vectors. Thanks!

Reply:
Mutable vector so I left it `std::vector`.